small functional improvements

robko23 opened the pull request:
# Summary

Hi, first of all - amazing work what you've done on the next branch. The api is absolutely amazing, but it still has some rough edges.
I've done one integration on 0.40 and now I'm building another integration and I've decided to use the next branch, so that I can test its ergonomics and help to improve it.

I would consider this PR a draft for now, because I'm just starting with the integration and I probably will send more patches if something doesn't work or doesn't seem right.

For now this PR contains two small commits - everything is explained in detail in the commit message as to why I've done it, but here is a list of things I've done (I will be editing this list with the new patches as I create them):

- export `StripeRequest` trait - allow `.customize()` to be used outside this crate
- change `Idempotency-Key` to `idempotency-key` - caused panic on runtime

### Checklist

- [x] ran `cargo make fmt`
- [x] using [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) to hightlight user-facing fixes and features

mzeitlin11 commented on async-stripe/src/hyper/client.rs:
```diff
@@ -109,7 +109,9 @@ impl Client {
 
         if let Some(key) = strategy.get_key() {
-            req_builder = req_builder.header(HeaderName::from_static("Idempotency-Key"), key);
+            const HEADER_NAME: HeaderName = HeaderName::from_static("idempotency-key");
```
Wow, that's a bad panic! Think this must've happened because the old usage of `http-types` instead of `hyper/http` supported this, and there must be no test coverage of idempotency key usage

mzeitlin11 commented on async-stripe/src/hyper/client.rs:
```diff
@@ -109,7 +109,9 @@ impl Client {
         if let Some(key) = strategy.get_key() {
-            req_builder = req_builder.header(HeaderName::from_static("Idempotency-Key"), key);
+            const HEADER_NAME: HeaderName = HeaderName::from_static("idempotency-key");
+            assert!(!key.is_empty(), "idempotency key is empty");
```
Probably would be better to return an error here, could add a variant to `StripeError`. Even better would be to validate on construction of `RequestStrategy`, but that's not currently possible without hiding more of the internal representation there.

robko23 replied:
yeah, I put that assertion here because I wanted to narrow down the out of bound index and then I left it there because it seemed a good sanity check.
However I also think that having a proper error would be better, so I created a commit that adds a newtype for the idempotency key. It is basically a string that cannot be empty or longer than 255 chars. It has its own error, because I think its mostly isolated from the rest of the logic. Having it mostly transparent struct also doesn't hide any implementation and representation

mzeitlin11 reviewed (commented):
Thanks for contributing this @robko23, there are definitely lots of improvements that can be made! 

I know this is in draft, but I'd be happy to merge this with the added `assert!` either changed to an error or removed since the other 2 changes should be pretty uncontroversial improvements.

robko23 commented:
I've removed the assert and if you want to I can mark it as ready and you can merge it, but there will be more small changes from me as I use it more. I didn't want to overwhelm you with these one line PRs so my idea was to pack it in a single PR with many small commits. But if you don't mind this, I can break it down to more PRs

mzeitlin11 commented:
Sorry for the delay here @robko23! I think the newtype approach is the right one generally, and I think that commit will be useful at a later point, but I'd imagine there will be more rounds of iteration on `RequestStrategy` going forward, so would like to hold off on breaking changes there since I think larger changes will come and don't want to have users testing out the `next` branch need to migrate their code too much.  

We can also make an issue to start tracking / discussing what the most useful form of `RequestStrategy` might be given the new structure of having custom clients in the `next` branch.

I'd be happy to merge if you don't mind just reverting the `RequestStrategy` changes and the original `assert`, so that this just fixes the bug.

robko23 commented:
Okay, so I cherry picked the commit with the newtype and I will create a new pull request with it. Then I removed this from this branch.

Then I changed the commit with the assert and removed it, but I still kept the `const HEADER_NAME`

mzeitlin11 reviewed (approved):
Thanks @robko23! CI failure should be unrelated, I have a fix in #660, which I will pull out separately.

mzeitlin11 commented:
Ok, sorry for the annoyance, but CI should be good now, so if you can rebase I'll merge it!

arlyon commented:
Hey thanks for this @robko23 ! I have been MIA for a little bit (was very ill for a couple of months the end of last year but I am back on form). @mzeitlin11 seems to be carrying the torch just fine in the mean time but I will stamp anyways.

arlyon commented:
Seems like there is a genuine clippy lint here. Feel free to @ either myself or Matthew when this is addressed

robko23 commented:
hi @arlyon! Thanks for the response and I hope your health will be better in the future. 

About the clippy lint, I have decided to ignore (`#[allow]`) it since it seems like it is false positive.
It took me a while to trigger the lint on my machine and I figured out you have older version of rustc on the CI. You are using 1.75.0, I have 1.84.1 and it does not trigger there.
The reason why it gets triggered on the older version is that `HeaderName` uses `bytes::Bytes` and that uses `AtomicPtr` which triggers the lint.
`bytes::Bytes` is by default ignored by the clippy lint and for some reason clippy doesn't figure out this relationship and triggers the lint.

See [this](https://github.com/rust-lang/rust-clippy/issues/5812), [this](https://github.com/hyperium/http/blob/a463fb5995d67e12b154474ede14a093db64f006/src/header/name.rs#L33-L51) and [this](https://rust-lang.github.io/rust-clippy/master/index.html#declare_interior_mutable_const) for more information

mzeitlin11 commented:
Sorry for the delay here, just running the CI now!

augustoccesar commented:
We can bump the MSRV, but these should also fix the CI:
- https://github.com/arlyon/async-stripe/pull/672
- https://github.com/arlyon/async-stripe/pull/673

augustoccesar commented:
> > We can bump the MSRV, but these should also fix the CI:
> > 
> > * [chore: bump upload-artifact to v4 #672](https://github.com/arlyon/async-stripe/pull/672)
> > * [fix: pin native-tls to <= 0.2.13 #673](https://github.com/arlyon/async-stripe/pull/673)
> 
> Pinning `native-tls` works too but just noting this is a PR to the `next` branch, so CI looks different and there is no upload-artifact.

Oh, sorry. I missed that this was on `next`! Disregard my comment!

arlyon commented:
The rust project have reversed their recommendations not to commit the Cargo.lock file so we may consider committing that to make CI less flaky.

arlyon commented:
Done! TY
